forge-ai/src/main/java/forge/ai/ability/EffectAi.java

```diff
@@ -211,9 +211,19 @@ public boolean apply(final Card c) {
                 if (game.getStack().isEmpty()) {
                     return false;
                 }
-                final SpellAbility topStack = game.getStack().peekAbility();
+                SpellAbility topStack = game.getStack().peekAbility();
                 final Player tgtPlayer = topStack.getActivatingPlayer();
-                return tgtPlayer.isOpponentOf(ai) && tgtPlayer.canGainLife() && topStack.getApi() == ApiType.GainLife;
+                if (tgtPlayer.isOpponentOf(ai) && tgtPlayer.canGainLife()) {
+                    while (topStack != null) {
+                        if (topStack.getApi() == ApiType.GainLife) {
+                            if ("You".equals(topStack.getParam("Defined")) || topStack.isTargeting(tgtPlayer)) {
+                                return true;
+                            }
+                        }
+                        topStack = topStack.getSubAbility();
+                    }
```

Comment:
would `findSubAbilityByType` be good enough for this? or is it not good for the corner cases?

you also could check for damage if topStack source has lifelink

Reply:
Ah, a good idea :) Also, checking for combat lifelink is a good idea too.
Not sure if findSubAbilityByType would help make it cleaner here tbh, since we also need to check the primary SA as well... Do you have an idea for how to write it in a clean enough manner checking both the main SA and the subs?